### PR TITLE
Re-enable four tests disabled because of an old OpenSSL bug

### PR DESCRIPTION
**Pull request by zhangsenWang**
## Description
Re-enable five tests disabled because of an old OpenSSL bug. Fixes https://github.com/Mbed-TLS/mbedtls/issues/5077

## Status
**READY**

## Steps to test or reproduce
with openssl 1.1.1a or higher version
`export OPENSSL_NEXT=/path/to/openssl`
`./tests/ssl-opt.sh -f "DTLS fragmenting: 3d, openssl"`
`./tests/ssl-opt.sh -f "DTLS proxy: 3d, openssl"`

### CODE REVIEW

**Comment by xkqian** [comment]
Seems DCO failed. You should pay attention to your commit message and add the signoff into it.

**Comment by zhangsenWang** [comment]
> Seems DCO failed. You should pay attention to your commit message and add the signoff into it.

I have already added the signoff into it.

**Comment by minosgalanakis** [comment]
Thank you for your contribution.

Would it be possible to reword the commit message so it describes what the patch does? Issue #5077 can be tracked through here.

Also does that mean that before merging this comit we need to update the CI images with the fixed version of OpenSSL?

**Comment by gilles-peskine-arm** [comment]
@minosgalanakis We already have a recent enough version of openssl on the CI images. (A CI pass confirms that.) The tests are from before we had this recent version and we hadn't gotten around to re-enabling them.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Patch: approved.

Changelog message: please write a meaningful changelog message. See `git log` for examples. The issue number is not meaningful in itself. The issue should always be linked in the pull request description, but it isn't vital to have it in the commit message since it can be traced back from the commit through the pull request. “Re-enable tests disabled because of an old OpenSSL bug” would be an adequate commit message (and pull request title).

This needs a backport to the `mbedtls-2.28` long-time support branch. Note that in 2.28, there are 7 test cases with `skip_next_test`.

**Comment by zhangsenWang** [comment]
> 

Thanks for your advice. I have changed the commit message. And I will do the backport with mbedtls-2.28 in another PR soon.

**Comment by xkqian** [comment]
We can see the comments below: 
"## Besides, openssl should use dtls1_2 or dtls, otherwise it will cause "SSL alert number 70" error"

Seems If we use the dtls1, the test will fail because of the version mismatch.  I check the code in server and find that we disable the dtls 1.0 support, is that the intention? @gilles-peskine-arm

**Comment by gilles-peskine-arm** [comment]
Mbed TLS 2.28 supports DTLS 1.0 and 1.2. Mbed TLS 3.x only supports 1.2.

OpenSSL 1.0.2g (`$OPENSSL`) only supports DTLS 1.0. OpenSSL 1.1.1a (`$OPENSSL_NEXT`) supports 1.0 and 1.2.

It should be ok to change these tests to `-dtls1_2` in 2.28, like you did in development.

**Review by gilles-peskine-arm — DISMISSED** [review]
Thanks for improving the commit message. Looks good to me now.

There's a CI failure now though. Please check what's going on.

File: tests/ssl-opt.sh
Comment (xkqian) [inline tests/ssl-opt.sh:9175]:
```suggestion

```

**Review by xkqian — CHANGES_REQUESTED** [review]

**Review by xkqian — APPROVED** [review]
LGTM

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
There's an inconsistency here. I think I know what's going on but this warrants further investigation. We need to make sure we understand what's going on.

File: tests/ssl-opt.sh
Comment (zhangsenWang) [reply tests/ssl-opt.sh]: Thanks for you suggestion, you are right. I have performed more tests with OpenSSL below 1.1.1a. The bug did only occur with OpenSSL server. I have deleted the version requirement for OpenSSL client. Please review again, thank you.

**Comment by xkqian** [comment]
Travis CI failed: 
`DTLS fragmenting: 3d, openssl client, DTLS 1.2 ......................... RETRY(client-timeout) RETRY(client-timeout) FAIL
  ! bad client exit code (expected 0, got 143)
  ! outputs saved to o-XXX-625.log`

**Comment by zhangsenWang** [comment]
> Travis CI failed: `DTLS fragmenting: 3d, openssl client, DTLS 1.2 ......................... RETRY(client-timeout) RETRY(client-timeout) FAIL ! bad client exit code (expected 0, got 143) ! outputs saved to o-XXX-625.log`

**Comment by xkqian** [comment]
Seems when `seed=1`, the pseudo random number will be the special one, which will cause the the test case time out failure(for examble drop 1:N packet, but the data size is multiplication of N). 
Suggest adding the option "seed" with different value from 1, such as `seed=2`
How about your opinion? @gilles-peskine-arm @zhangsenWang

**Comment by zhangsenWang** [comment]
With certain seed, the DTLS test with OpenSSL client will time out. I find that this is because of an OpenSSL bug that the client will get stuck in case of duplicated  `Encrypted handshake message`,  I report it in https://github.com/openssl/openssl/issues/18887. As for the test cases using OpenSSL server, they passed according to the log in Jenkins.

**Comment by xkqian** [comment]
Please skip the failure cases and go on this PR.  I think we can file one new issue with description: "With certain seed, the DTLS test with OpenSSL client will time out".

**Comment by xkqian** [comment]
Something wrong with CI.  Please remove the tail white space.
image

**Review by xkqian — APPROVED** [review]
LGTM

**Review by gilles-peskine-arm — APPROVED** [review]

**Pull request merged**
